base/applications/msconfig/autochkpage.c

```diff
@@ -0,0 +1,247 @@
+                                   0,
+                                   REG_MULTI_SZ,
+                                   (const BYTE *)EnableValue,
+                                   sizeof(Program));
+
+            if (Status == ERROR_SUCCESS)
+            {
+              // Value set has completed successfully, autochk will be enabled
+              SendDlgItemMessage(hDlg, IDC_AUTOCHK_ENABLE, BM_SETCHECK, BST_CHECKED, 0);
+              RegCloseKey(hKey);
+            }
+            else
+            {
+              // We failed, close the key
+              RegCloseKey(hKey);
+              return 1;
+            }
+          }
+          else
+          {
+            // We failed, close the key
+            RegCloseKey(hKey);
+            return 1;
+          }
+          
+          break;
+        }
+
+        case IDC_AUTOCHK_DISABLE:
+        {
+
+          // Open the key so we can query it
+          Status = RegOpenKeyEx(HKEY_LOCAL_MACHINE,
+                                SubKey,
+                                0,
+                                KEY_ALL_ACCESS,
+                                &hKey);
+
+          if (Status == ERROR_SUCCESS)
+          {
+
+            Status = RegQueryValueExW(hKey,
+                                      ValueName,
+                                      NULL,
+                                      &dwType,
+                                      NULL,
+                                      &dwcbData);
+
+            if (Status != ERROR_SUCCESS || dwType != REG_MULTI_SZ)
+            {
+              // Error, bail out
+              RegCloseKey(hKey);
+              return 1;
+            }
+
+            // Allocate a heap for the variable
+            lpsszMultiString = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, dwcbData);
+
+            Status = RegQueryValueExW(hKey,
+                                      ValueName,
+                                      NULL,
+                                      NULL,
+                                      (LPBYTE)lpsszMultiString,
+                                      &dwcbData);
+
+            if (Status != ERROR_SUCCESS)
+            {
+              // An error happened... free memory & close the key & bail out
+              HeapFree(GetProcessHeap(), 0, lpsszMultiString);
+              RegCloseKey(hKey);
+              return 1;
+            }
+
+            dwBufferSize = dwcbData;
+            ptr = lpsszMultiString;
+
+            while (*ptr)
+            {
+              szSize = wcslen(ptr) + 1;
+              if (_wcsnicmp(ptr, L"autocheck ", 10) == 0)
+              {
+                /*
+                * Found a string to remove. To do that, we just
+                * move everything that is after it backwards.
+                */
+                dwBufferSize -= szSize * sizeof(WCHAR); // Size that remains after the string to be removed.
+                memmove(ptr, ptr + szSize, dwBufferSize);
```

Comment:
`dwBufferSize` will be too large here, unless we're at the beginning of the string.
Example: `mycommand\0autocheck autochk *\0othercommand\0\0`
`dwBufferSize` would be `wcslen("mycommand")+1 + wcslen("othercommand")+1 + 1`, whereas you only want to copy `othercommand\0\0`

Reply:
We're at the beginning of the string. The actual snippet of code is provided by Hermes.

Reply:
Actually @ThFabba is right, the code snippet had a bug.

Reply:
Therefore in the 202 line I'll just replace with `wcslen("mycommand")+1 + wcslen("othercommand")+1 + 1`?  And `mycommand` will be `autocheck` and `othercommand` will be `autochk *`?

Reply:
9 1 19 1 12 1 11 1 1

What you want to do is this, right?
`mycommand\0autocheck autochk *\0othercommand\0nextcommand\0\0` -> `mycommand\0othercommand\0nextcommand\0\0`
`dwBufferSize == 112`

But what will happen is:
* first iteration, `ptr == L"mycommand"`, `szSize == 10`; then we adjust ptr by 10
* next iteration, `ptr == L"autocheck autochk *"`, `szSize == 20`; we go into the if
** dwBufferSize gets reduced from 112 to 92
** memmove from "othercommand..." to "autocheck ..." with len=92: the entire string becomes `L"mycommand\0othercommand\0nextcommand\0\0<20 chars from uninitialized memory>"`. The correct length would have been 56, not 92
* next iteration, `ptr = L"othercommand"`... the rest is boring and works as expected